batch2/batch/batch.py

```diff
@@ -2,450 +2,244 @@
-        self.deleted = True
-        self.closed = True
-        log.info(f'batch {self.id} marked for deletion')
-
-    async def delete(self):
-        # Job deleted from database when batch is deleted with delete cascade
-        for j in await self.get_jobs():
-            await j._delete_gs_files()
-        await self.app['db'].batch.delete_record(self.id)
-        log.info(f'batch {self.id} deleted')
-
-    async def mark_job_complete(self):
-        if self.complete and self.callback:
-            log.info(f'making callback for batch {self.id}: {self.callback}')
-            try:
-                async with aiohttp.ClientSession(
-                        raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
-                    await session.post(self.callback, json=await self.to_dict(include_jobs=False))
-                    log.info(f'callback for batch {self.id} successful')
-            except Exception:  # pylint: disable=broad-except
-                log.exception(f'callback for batch {self.id} failed, will not retry.')
-
-    def is_complete(self):
-        return self.complete
-
-    def is_successful(self):
-        return self.state == 'success'
-
-    async def to_dict(self, include_jobs=False, limit=None, offset=None):
-        result = {
-            'id': self.id,
-            'state': self.state,
-            'complete': self.complete,
-            'closed': self.closed
-        }
-        if self.attributes:
-            result['attributes'] = self.attributes
-        if include_jobs:
-            jobs = await self.get_jobs(limit, offset)
-            result['jobs'] = sorted([j.to_dict() for j in jobs], key=lambda j: j['job_id'])
-        return result
+    job_spec = json.loads(record['spec'])
+
+    secrets = job_spec['secrets']
+    k8s_secrets = await asyncio.gather(*[
+        k8s_client.read_namespaced_secret(
+            secret['name'], BATCH_NAMESPACE,
+            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
+        for secret in secrets
+    ])
+    for secret, k8s_secret in zip(secrets, k8s_secrets):
+        secret['data'] = k8s_secret.data
+
+    return {
+        'batch_id': record['batch_id'],
+        'user': record['user'],
+        'job_spec': job_spec
+    }
+
+
+async def schedule_job(app, record, instance):
+    assert instance.state == 'active'
+
+    db = app['db']
+
+    batch_id = record['batch_id']
+    job_id = record['job_id']
+    id = (batch_id, job_id)
+
+    async with aiohttp.ClientSession(
+            raise_for_status=True, timeout=aiohttp.ClientTimeout(total=60)) as session:
+        url = f'http://{instance.ip_address}:5000/api/v1alpha/batches/jobs/create'
+        body = await job_config(app, record)
+        await request_retry_transient_errors(
+            session, 'POST',
+            url, json=body)
+
+    log.info(f'schedule job {id} on {instance}: called create job')
+
+    await check_call_procedure(
```

Comment:
See above comment; they are caught in Scheduler.loop and the scheduler loop tries again.

If the database goes down, we could schedule something multiple times in the course of one batch.  I see a few options:
 - Do nothing, the database won't normally be down.  Just let it overschedule sometimes.
 - Let per-instance healing fix it (driver sends a list of jobs expected to be running on an instance with requests, instance sends its running jobs with responses).  This will potentially generate a lot of database traffic.  Maybe we just do it sometimes.  I think we'll eventually want something like this.
 - If an exception gets thrown, try to cancel the just we just submitted.  If that fails, give up.  You still get schedule-unschedule-reschedule loops when the database goes down.

I'm fine with the first one for now.  It will only double-schedule with in a single (50 job) scheduling batch, because when it goes back to get more pods, the database will be unavailable.

Note, I plan to retry database timeout errors (I consider those transient errors), so actually we'll just keep retrying the call.  What exception are we worried about here?

Reply:
I was more concerned that we'd schedule the job and then batch2 would get redeployed and the fact that it was scheduled on the instance is lost. If that were to happen, then the job status would get sent back to the driver from an instance that was unexpected if we schedule again on a different instance or if it's the same instance, the worker should throw an exception saying it already has that job. But then what happens? My mental model of the exceptions is saying that this will get caught in an infinite loop of trying to reschedule if we have the same instance, but I'm not sure.